modules/glfw/src/filewatcher.cpp

```diff
@@ -0,0 +1,325 @@
+                changed.emplace_back(elem, Action::Removed);
+                files[elem] = filesystem::fileModificationTime(elem);
+            } else {
+                auto newTime = filesystem::fileModificationTime(elem);
+                if (newTime > it->second) {
+                    changed.emplace_back(elem, Action::Modified);
+                    it->second = newTime;
+                }
+            }
+        }
+
+        return changed;
+    }
+
+    static constexpr DWORD filter =
+        FILE_NOTIFY_CHANGE_FILE_NAME | FILE_NOTIFY_CHANGE_FILE_NAME | FILE_NOTIFY_CHANGE_LAST_WRITE;
+
+    std::array<HANDLE, MAXIMUM_WAIT_OBJECTS> handles_{};
+    std::array<std::pair<std::string, std::unordered_map<std::string, time_t>>,
+               MAXIMUM_WAIT_OBJECTS>
+        observed_{};
+    std::atomic<size_t> active_ = 0;
+
+    std::function<void(const std::string&, const std::string&, Action)> changeCallback_;
+    std::mutex mutex_;
+    std::vector<std::string> toAdd_;
+    std::vector<std::string> toRemove_;
+    std::atomic<bool> stop_{false};
+    std::chrono::milliseconds timeout_{1000};
+    std::thread thread_{[this]() { watch(); }};
+};
+
+FileWatcher::FileWatcher(InviwoApplication* app)
+    : app_{app}
+    , watcher_{std::make_unique<WatcherThread>(
+          [this](const std::string& dir, const std::string& path, WatcherThread::Action) {
+              auto notifyAboutChanges = [this, dir, path]() {
+                  if (filesystem::fileExists(path)) {
+                      // don't use iterators here, they might be invalidated.
+                      const auto orgSize = fileObservers_.size();
+                      for (size_t i = 0; i < orgSize && i < fileObservers_.size(); ++i) {
+                          if (fileObservers_[i]->isObserved(path)) {
+                              fileObservers_[i]->fileChanged(path);
+                          }
+                      }
+                  }
+                  if (filesystem::directoryExists(dir)) {
+                      // don't use iterators here, they might be invalidated.
+                      const auto orgSize = fileObservers_.size();
+                      for (size_t i = 0; i < orgSize && i < fileObservers_.size(); ++i) {
+                          if (fileObservers_[i]->isObserved(dir)) {
+                              fileObservers_[i]->fileChanged(dir);
+                          }
+                      }
+                  }
+              };
+
+              if (app_) {
+                  app_->dispatchFront(notifyAboutChanges);
+              } else {
+                  notifyAboutChanges();
+              }
+          })} {}
+
+FileWatcher::~FileWatcher() = default;
+
+void FileWatcher::registerFileObserver(FileObserver* fileObserver) {
+    IVW_ASSERT(std::find(fileObservers_.cbegin(), fileObservers_.cend(), fileObserver) ==
+                   fileObservers_.cend(),
+               "File observer already registered.");
```

Comment:
Is it a good idea to use only an assertion here?

Reply:
that is handled internally by the FileObserver hence it should not really happen by user error. So I think the assert is good.